cpp/SnowflakeS3Client.cpp

```diff
@@ -98,6 +121,126 @@ TransferOutcome SnowflakeS3Client::upload(FileMetadata *fileMetadata,
   }
 }
 
+void *Snowflake::Client::SnowflakeS3Client::uploadParts(MultiUploadCtx * uploadCtx)
+{
+  Aws::S3::Model::UploadPartRequest uploadPartRequest;
+
+  uploadPartRequest.WithBucket(uploadCtx->m_bucket)
+                   .WithKey(uploadCtx->m_key);
+
+  uploadPartRequest.SetContentType(CONTENT_TYPE_OCTET_STREAM);
+  uploadPartRequest.SetContentLength(uploadCtx->buf->getSize());
+  uploadPartRequest.SetBody(Aws::MakeShared<Aws::IOStream>("", uploadCtx->buf));
+  uploadPartRequest.SetUploadId(uploadCtx->m_uploadId);
+  uploadPartRequest.SetPartNumber(uploadCtx->m_partNumber);
+
+  Aws::S3::Model::UploadPartOutcome outcome = s3Client->UploadPart(uploadPartRequest);
+
+  if (outcome.IsSuccess())
+  {
+    uploadCtx->m_outcome = TransferOutcome::SUCCESS;
+    uploadCtx->m_etag = outcome.GetResult().GetETag();
+  }
+  else
+  {
+    fprintf(stderr, "Upload request: %s\n", outcome.GetError().GetMessage().c_str());
+    uploadCtx->m_outcome = TransferOutcome::FAILED;
+  }
+}
+
+TransferOutcome SnowflakeS3Client::doMultiPartUpload(FileMetadata *fileMetadata,
+  std::basic_iostream<char> *dataStream)
+{
+  if (m_threadPool == nullptr)
+  {
+    m_threadPool = new Util::ThreadPool(m_parallel);
+  }
+
+  std::map<std::string, std::string> userMetadata;
+  addUserMetadata(&userMetadata, fileMetadata);
+
+  std::string bucket, key;
+  extractBucketAndKey(fileMetadata, bucket, key);
+
+  Aws::S3::Model::CreateMultipartUploadRequest request;
+  request.WithBucket(bucket)
+    .WithKey(key)
+    .WithContentType(CONTENT_TYPE_OCTET_STREAM)
+    .WithMetadata(userMetadata) ;
+
+  auto createMultiPartResp = s3Client->CreateMultipartUpload(request);
+  if (createMultiPartResp.IsSuccess())
+  {
+    Aws::String uploadId = createMultiPartResp.GetResult().GetUploadId();
+
+    Util::StreamSplitter splitter(dataStream, m_parallel, DATA_SIZE_THRESHOLD);
+    unsigned int totalParts = splitter.getTotalParts(
+      fileMetadata->encryptionMetadata.cipherStreamSize);
+
+    MultiUploadCtx uploadParts[totalParts];
+
+    for (unsigned int i = 0; i < totalParts; i++)
+    {
+      uploadParts[i].m_uploadId = uploadId;
+      uploadParts[i].m_partNumber = i+1;
+      uploadParts[i].m_bucket = bucket;
+      uploadParts[i].m_key = key;
+
+      m_threadPool->AddJob([&splitter, i, this, &uploadParts]()->void
+                           {
+                             Util::ByteArrayStreamBuf * buf = splitter.getNextSplitPart();
+                             uploadParts[i].buf = buf;
+                             this->uploadParts(&uploadParts[i]);
+                             splitter.markDone(buf);
+                           });
+    }
+
+    m_threadPool->WaitAll();
```

Comment:
could this raise any known exception? such as ExpiredToken?

Reply:
Expired token is not supported currently. I probably do it in separate pr. This change is already big enough.